### PR TITLE
Pull in recent PWM updates from Zephyr, align accordingly boards, apps and samples

### PR DESCRIPTION
**Pull request by anangl**
See descriptions of particular commits for more detailed information.

### CODE REVIEW

**Review by gmarull — APPROVED** [review]
lgtm, thanks for these changes!

**Comment by mbolivar-nordic** [comment]
@anangl integration is failing

**Comment by zycz** [comment]
> Hi @jumi-nordic or @zycz , can you quickly check on Monday/Tuesday if leds are working as expected? For example on gmouse.

Checked out on this PR and checked leds on gmouse. Everything seams to work fine.

**Comment by anangl** [comment]
Rebased.

File: doc/nrf/libraries/caf/leds.rst [hunk 1]
```diff
@@ -224,13 +224,13 @@ The following code snippets show examples of the DTS nodes:
...
 			status = "okay";
-			pwms = <&pwm1 4>;
+			pwms = <&pwm1 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
 			label = "LED1";
 		};
 	};
 
      In this example, ``pwmleds0`` is a tri-channel color LED node, while ``pwmleds1`` is a monochromatic LED node.
-     Both ``pwmleds`` nodes are pointing to the ``pwms`` properties corresponding to PWM nodes set in Example 2 in `Enabling the PWM ports`_, with the respective channel GPIO pin numbers.
+     Both ``pwmleds`` nodes are pointing to the ``pwms`` properties corresponding to PWM nodes set in Example 2 in `Enabling the PWM ports`_, with the respective channel numbers.
```
Comment (umapraseeda) [inline doc/nrf/libraries/caf/leds.rst:233]: These two sentences are appearing inside the code block. Is it supposed to be like that? (See the 'Example 2' section in http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-7597/nrf/libraries/caf/leds.html#enabling-the-led-pwm-nodes)

Comment (anangl) [reply doc/nrf/libraries/caf/leds.rst:233]: Right. I corrected the indentation.

**Review by umapraseeda — APPROVED** [review]

**Review by pdunaj — APPROVED** [review]

**Review by KAGA164 — APPROVED** [review]
I checked the NFC Reader(st25r3911b), it works fine with that changes.

**Comment by thst-nordic** [comment]
This passed all integration testing,  what is left to merge?

File: samples/bluetooth/mesh/light_ctrl/src/lc_pwm_led.c [hunk 2]
```diff
@@ -49,9 +33,5 @@ void lc_pwm_led_set(uint16_t desired_lvl)
 		(PWM_PERIOD * desired_lvl) /
```
Comment (Damian-Nordic) [inline samples/bluetooth/mesh/light_ctrl/src/lc_pwm_led.c:33]: Should we use `PWM_PERIOD` here or rather take it from the DTS spec?

File: samples/bluetooth/nrf_dm/src/pwm_led.c [hunk 3]
```diff
@@ -50,5 +33,5 @@ void pwm_led_set(uint16_t desired_lvl)
 {
 	uint32_t scaled_lvl = (PWM_PERIOD * desired_lvl) / LIGHTNESS_MAX;
```
Comment (Damian-Nordic) [inline samples/bluetooth/nrf_dm/src/pwm_led.c:34]: Is `PWM_PERIOD` aligned with the DTS?

File: samples/matter/common/src/pwm_device.cpp [hunk 4]
```diff
@@ -102,5 +102,6 @@ void PWMDevice::UpdateLight()
 	const uint8_t effectiveLevel =
 		mState == kState_On ? chip::min<uint8_t>(mLevel - mMinLevel, maxEffectiveLevel) : 0;
 
-	pwm_pin_set_usec(mPwmDevice, mPwmChannel, kPwmWidthUs, kPwmWidthUs * effectiveLevel / maxEffectiveLevel, 0);
+	pwm_set(mPwmDevice, mPwmChannel, PWM_USEC(kPwmWidthUs),
+		PWM_MSEC(kPwmWidthUs * effectiveLevel / maxEffectiveLevel), 0);
```
Comment (Damian-Nordic) [inline samples/matter/common/src/pwm_device.cpp:106]:
```suggestion
		PWM_USEC(kPwmWidthUs * effectiveLevel / maxEffectiveLevel), 0);
```

File: samples/nrf9160/lwm2m_client/src/ui/ui_buzzer.c [hunk 5]
```diff
@@ -59,12 +57,12 @@ int ui_buzzer_on_off(bool new_state)
...
 	if (frequency == 0) {
 		/* Turn off buzzer when frequency = 0 */
-		ret = pwm_pin_set_usec(buzzer_dev, buzzer_pin, USEC_PER_SEC, 0, 0);
+		ret = pwm_set_pulse_dt(&buzzer, 0);
 	} else {
 		uint32_t period = PERIOD(frequency);
 		uint32_t pulse_width = calculate_pulse_width(period, intensity);
 
-		ret = pwm_pin_set_usec(buzzer_dev, buzzer_pin, period, pulse_width * state, 0);
+		ret = pwm_set_dt(&buzzer, PWM_USEC(period), pulse_width * state);
```
Comment (Damian-Nordic) [inline samples/nrf9160/lwm2m_client/src/ui/ui_buzzer.c:65]: If `period` is USEC, should the last argument be also wrapped with `PWM_USEC`?

File: samples/nrf9160/lwm2m_client/src/ui/ui_led.c [hunk 6]
```diff
@@ -59,10 +41,9 @@ int ui_led_pwm_on_off(uint8_t led_num, bool new_state)
 
 	state[led_num] = new_state;
 
-	ret = pwm_pin_set_usec(pwm_leds[led_num].dev, pwm_leds[led_num].channel,
-			       PWM_PERIOD_USEC,
-			       pulse_width[led_num] * new_state,
-			       pwm_leds[led_num].flags);
+	ret = pwm_set_dt(&pwm_leds[led_num],
+			 PWM_USEC(PWM_PERIOD_USEC),
+			 pulse_width[led_num] * new_state);
```
Comment (Damian-Nordic) [inline samples/nrf9160/lwm2m_client/src/ui/ui_led.c:46]: what units `pulse_width[led_num]` is in? Should we add `PWM_USEC`, too?

File: samples/nrf9160/lwm2m_client/src/ui/ui_led.c [hunk 7]
```diff
@@ -86,10 +67,9 @@ int ui_led_pwm_set_intensity(uint8_t led_num, uint8_t led_intensity)
 
 	pulse_width[led_num] = calculate_pulse_width(led_intensity);
 
-	ret = pwm_pin_set_usec(pwm_leds[led_num].dev, pwm_leds[led_num].channel,
-			       PWM_PERIOD_USEC,
-			       pulse_width[led_num] * state[led_num],
-			       pwm_leds[led_num].flags);
+	ret = pwm_set_dt(&pwm_leds[led_num],
+			 PWM_USEC(PWM_PERIOD_USEC),
+			 pulse_width[led_num] * state[led_num]);
```
Comment (Damian-Nordic) [inline samples/nrf9160/lwm2m_client/src/ui/ui_led.c:72]: ditto

File: samples/zigbee/light_bulb/src/main.c [hunk 8]
```diff
@@ -287,8 +273,7 @@ static void light_bulb_set_brightness(zb_uint8_t brightness_level)
 {
 	uint32_t pulse = brightness_level * LED_PWM_PERIOD_US / 255U;
 
-	if (pwm_pin_set_usec(led_pwm_dev, PWM_DK_LED4_CHANNEL,
-			     LED_PWM_PERIOD_US, pulse, PWM_DK_LED4_FLAGS)) {
+	if (pwm_set_dt(&led_pwm, PWM_USEC(LED_PWM_PERIOD_US), pulse)) {
```
Comment (Damian-Nordic) [inline samples/zigbee/light_bulb/src/main.c:276]:
```suggestion
	if (pwm_set_dt(&led_pwm, PWM_USEC(LED_PWM_PERIOD_US), PWM_USEC(pulse))) {
```

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]

File: samples/bluetooth/mesh/light_ctrl/src/lc_pwm_led.c (hunk 2 above)
Comment (anangl) [reply samples/bluetooth/mesh/light_ctrl/src/lc_pwm_led.c:33]: I replaced the call below with `pwm_set_dt` and used `PWM_PERIOD` there.

File: samples/bluetooth/nrf_dm/src/pwm_led.c (hunk 3 above)
Comment (anangl) [reply samples/bluetooth/nrf_dm/src/pwm_led.c:34]: I replaced the call below with `pwm_set_dt` and used `PWM_PERIOD` there.

File: samples/matter/common/src/pwm_device.cpp (hunk 4 above)
Comment (anangl) [reply samples/matter/common/src/pwm_device.cpp]: Fixed.

File: samples/nrf9160/lwm2m_client/src/ui/ui_buzzer.c (hunk 5 above)
Comment (anangl) [reply samples/nrf9160/lwm2m_client/src/ui/ui_buzzer.c]: Yes, fixed.

File: samples/nrf9160/lwm2m_client/src/ui/ui_led.c (hunk 6 above)
Comment (anangl) [reply samples/nrf9160/lwm2m_client/src/ui/ui_led.c]: Yes, wrapped in `PWM_USEC`.

File: samples/nrf9160/lwm2m_client/src/ui/ui_led.c (hunk 7 above)
Comment (anangl) [reply samples/nrf9160/lwm2m_client/src/ui/ui_led.c]: Fixed.

File: samples/zigbee/light_bulb/src/main.c (hunk 8 above)
Comment (anangl) [reply samples/zigbee/light_bulb/src/main.c]: Fixed.

**Comment by anangl** [comment]
@Damian-Nordic Thanks for the review! I addressed your comments.

File: samples/nrf9160/lwm2m_client/boards/thingy91_nrf9160_ns.overlay [hunk 9]
```diff
@@ -7,6 +7,6 @@
 / {
 	buzzer: buzzer {
 		compatible = "pwm-buzzer";
-		pwms = <&pwm1 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
+		pwms = <&pwm1 0 PWM_HZ(440) PWM_POLARITY_NORMAL>;
```
Comment (Damian-Nordic) [inline samples/nrf9160/lwm2m_client/boards/thingy91_nrf9160_ns.overlay:10]: Good, I like songs in A key ;)

File: subsys/zigbee/osif/zb_nrf_led_button.c [hunk 10]
```diff
@@ -170,12 +164,7 @@ void zb_osif_led_on_set_level(zb_uint8_t level)
 {
 	uint32_t pulse = level * LED_PWM_PERIOD_US / 255U;
 
-	if (!led_pwm_dev) {
-		return;
-	}
-
-	if (pwm_pin_set_usec(led_pwm_dev, pwm_channel, LED_PWM_PERIOD_US,
-			     pulse, pwm_flags)) {
+	if (pwm_set_dt(led_pwm, PWM_USEC(LED_PWM_PERIOD_US), pulse)) {
```
Comment (Damian-Nordic) [inline subsys/zigbee/osif/zb_nrf_led_button.c:167]: Last occurrence:
```suggestion
	if (pwm_set_dt(led_pwm, PWM_USEC(LED_PWM_PERIOD_US), PWM_USEC(pulse))) {
```

**Review by Damian-Nordic — APPROVED** [review]
Found last issue. Also @ArekBalysNordic is going to verify Matter samples empirically.

**Comment by anangl** [comment]
Rebased.

File: samples/matter/light_bulb/boards/nrf21540dk_nrf52840.overlay [hunk 11]
```diff
@@ -27,7 +27,7 @@
 
 	pwmleds {
 		pwm_led1: pwm_led_1 {
-			pwms = < &pwm0 0xe >;
+			pwms = <&pwm0 1 PWM_MSEC(20) PWM_POLARITY_INVERTED>;
```
Comment (ArekBalysNordic) [inline samples/matter/light_bulb/boards/nrf21540dk_nrf52840.overlay:30]: I verified Matter samples empirically, and I found that the pwm_led logic had been changed. When I'm sending the command to switch the LED on, the PWM is set to 0; after the command to switch the LED off, the PWM is set to the maximum value.

**Review by ArekBalysNordic — CHANGES_REQUESTED** [review]

**Review by tomchy — APPROVED** [review]

File: samples/matter/light_bulb/boards/nrf21540dk_nrf52840.overlay (hunk 11 above)
Comment (anangl) [reply samples/matter/light_bulb/boards/nrf21540dk_nrf52840.overlay:30]: Fixed.

File: lib/st25r3911b/st25r3911b_spi.c [hunk 12]
```diff
@@ -62,31 +61,19 @@ static bool cmd_is_valid(uint8_t cmd)
...
 
-	if (err) {
-		return err;
+		return -ENXIO;
 	}
-	spi_dev = device_get_binding(SPI_BUS);
-	if (!spi_dev) {
-		LOG_ERR("Cannot find SPI device %s!", SPI_BUS);
+
+	if (!device_is_ready(spi_dev)) {
```
Comment (gmarull) [inline lib/st25r3911b/st25r3911b_spi.c:75]: better use `spi_is_ready()`

**Pull request merged**